File: 1770-maximum-score-from-performing-multiplication-operations/1770-maximum-score-from-performing-multiplication-operations.java
class Solution {
    public int maximumScore(int[] nums, int[] multipliers) {
        return dp(nums, multipliers, 0, 0, new int[nums.length][multipliers.length]);        
    }
    
    private int dp(int[] nums, int[] multipliers, int i, int left, int[][] memo) {
        if (i == multipliers.length)
            return 0;
        
        int right = nums.length - 1 - (i - left);
        
        if (memo[left][i] == 0) {
            memo[left][i] = Math.max(nums[left] * multipliers[i] + dp(nums, multipliers, i + 1, left + 1, memo), nums[right] * multipliers[i] + dp(nums, multipliers, i + 1, left, memo));                
        }
        
        return memo[left][i];
    }
}

File: 1770-maximum-score-from-performing-multiplication-operations/1770-maximum-score-from-performing-multiplication-operations.py
class Solution:
    def maximumScore(self, nums: List[int], multipliers: List[int]) -> int:
        if len(multipliers) == 0:
            return 0

        memo = {}
        def dp(i, left):
            if i >= len(multipliers):
                return 0
            right = len(nums) - 1 - (i - left)
            if (i, left) in memo:
                return memo[(i, left)]
            else:
                memo[(i, left)] = max(dp(i + 1, left + 1) + multipliers[i] * nums[left]
                                      , dp(i + 1, left) + multipliers[i] * nums[right])
                return memo[(i, left)]
        
        return dp(0, 0)